dask_mpi/core.py

```diff
@@ -0,0 +1,31 @@
+import sys
+
+from mpi4py import MPI
+from tornado.ioloop import IOLoop
+
+from dask_mpi.common import get_host_from_interface, start_scheduler, start_worker
+
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+loop = IOLoop()
+
+
+def initialize(scheduler_file='scheduler.json', interface=None, nthreads=1, local_directory='',
+               memory_limit='auto', scheduler=True, nanny=False, bokeh_port=8787, bokeh_prefix=None,
+               bokeh_worker_port=8789):
+    host = get_host_from_interface(interface)
+
+    if rank == 0 and scheduler:
+        start_scheduler(loop, host=host, scheduler_file=scheduler_file,
+                        bokeh_port=bokeh_port, bokeh_prefix=bokeh_prefix)
+        sys.exit()
+
+    elif rank == 1 or (rank == 0 and not scheduler):
+        pass
```

Comment:
I think that ideally we would place the `scheduler_file` argument into configuration in such a way that a naive `Client()` call would pick it up.  I wonder if it makes sense to add an entry to dask.distributed's config like 

```python
client:
    initialization-keywords: {}
```

That we can fill with entries like `dask.config.set({'distributed.client.initialization-kwargs.scheduler_file': scheduler_file})` or something similar.

Reply:
I think that's a nice idea.  It makes the look cleaner.

I also would be interested in having MPI broadcast the scheduler address, with something like:

```python
from mpi4py import MPI
comm = MPI.COMM_WORLD
scheduler_address = comm.Bcast(scheduler_address, root=0)
```

After the scheduler is able to obtain it's address.  This would block on the return from the `initialize` function, and it would block the creation of the workers.  

I've thought about putting this logic into something like an `MPICluster` object, so you could initialize with:

```python
from dask_mpi import MPICluster
cluster = MPICluster(...initialize args...)

from distributed import Client
client = Client(cluster)
```

That would match the "well known and approved" syntax.

Reply:
Incidentally, is this Client `config` capability working now?  Or would this be something to do in the future of `distributed`?

Reply:
The config capability is aspirational.  It is not currently in master (though could be placed there rather easily). 

Your idea of using MPI to communicate the scheduler address actually seems cleaner to me though.  I say that we do that.  Presumably in core.py all of the ranks would call recv, and then either start a worker or insert that value into the configuration and then return control back to the client script.

```python
elif rank == 1:  # client
    scheduler_address = mpi.recv(...)
    dask.config.set(scheduler_address=scheduler_address)
    return
else:  # workers
    scheduler_address = mpi.recv(...)
    start_worker(..., scheduler_address=scheduler_address)
```

This would remove the need for us to use `scheduler_file` parameter.

Reply:
Note that the use of the scheduler_address config parameter *is* in latest release and doesn't need special attention.

Reply:
Nice!  Ok.  I'll implement the MPI broadcast option.

Reply:
👍

Reply:
The `scheduler_file` parameter has been removed from the `initialize` method with commit https://github.com/dask/dask-mpi/pull/6/commits/fd335be0bfab4ec67cfe58b749bc3b2facbcd221.  This also implements the MPI broadcast call to share the scheduler address and the use of dask.config.